### PR TITLE
Added Descope as an OAuth Provider

### PR DESCRIPTION
**Pull request by gaokevin1**
# Why

Many of our customers have migrated away from Cognito and other authentication services and are looking for documentation for how to use Expo with Descope as an OIDC provider. This documentation should exist on Expo's website, as well as our external documentation as well.

# How

This is using `AuthSession` like a generic OAuth provider configured with Expo, however this also showcases how to exchange the OAuth code for the token using PKCE, as many of our customers have been asking for this.

# Test Plan

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by gaokevin1** [comment]
How do I also include the company logo as well, under `ASSETS`?

**Comment by Simek** [comment]
Hello @gaokevin1, thank you for adding an another Auto guide! 👍 

You need to add the entry to [docs/ui/components/Authentication/index.ts](https://github.com/expo/expo/blob/2a96b86e780a2cc6c4cf4d5b38849f108f0517b0/docs/ui/components/Authentication/index.ts) file, and place the image asset in [docs/public/static/images/sdk/auth-session](https://github.com/expo/expo/tree/2a96b86e780a2cc6c4cf4d5b38849f108f0517b0/docs/public/static/images/sdk/auth-session) directory.

**Comment by gaokevin1** [comment]
@Simek I added the asset files accordingly, thanks for your help! Could you also make sure that this gets merged as well: https://github.com/expo/examples/pull/446?

File: docs/pages/guides/authentication.mdx [hunk 1]
```diff
@@ -770,6 +771,137 @@ function useMounted() {
...
+          /* @end */
+          title="Login"
+          /* @info Prompt the user to authenticate in a user interaction or web browsers will block it. */
+          onPress={() => promptAsync()}
+          /* @end */
+        />
+      )}
+    </View>
+  );
+}
```
Comment (Simek) [inline docs/pages/guides/authentication.mdx:894]:
```suggestion
import * as React from 'react';
import * as WebBrowser from 'expo-web-browser';
import * as AuthSession from 'expo-auth-session';
import { Button, View } from 'react-native';

WebBrowser.maybeCompleteAuthSession();

const descopeProjectId = '<Descope Project ID>'; // Replace this with your Descope Project ID
const descopeUrl = `https://api.descope.com/${descopeProjectId}`;
const redirectUri = AuthSession.makeRedirectUri();

export default function App() {
  const [authTokens, setAuthTokens] = React.useState(null);
  const discovery = AuthSession.useAutoDiscovery(descopeUrl);

  const [request, response, promptAsync] = AuthSession.useAuthRequest(
    {
      clientId: descopeProjectId,
      responseType: AuthSession.ResponseType.Code,
      redirectUri,
      usePKCE: true,
      scopes: ['openid', 'profile', 'email'],
    },
    discovery
  );

  React.useEffect(() => {
    if (response) {
      if (response.error) {
        console.error(
          'Authentication error',
          response.params.error_description || 'something went wrong'
        );
        return;
      }
      if (response.type === 'success') {
        /* @info Exchange the code for an access token in a server. */
        const exchangeFn = async (exchangeTokenReq) => {
        /* @end */
          try {
            const exchangeTokenResponse = await AuthSession.exchangeCodeAsync(
              exchangeTokenReq,
              discovery
            );
            setAuthTokens(exchangeTokenResponse);
          } catch (error) {
            console.error(error);
          }
        };

        exchangeFn({
          clientId: descopeProjectId,
          code: response.params.code,
          redirectUri,
          extraParams: {
            code_verifier: request.codeVerifier,
          },
        });
      }
    }
  }, [discovery, request, response]);

  const logout = async () => {
    const revokeResponse = await AuthSession.revokeAsync(
      {
        clientId: descopeProjectId,
        token: authTokens.refreshToken,
      },
      discovery
    );
    if (revokeResponse) {
      setAuthTokens(null);
    }
  };

  return (
    <View>
      {authTokens ? (
        <Button title="Logout" onPress={logout} />
      ) : (
        <Button
          /* @info Disable the button until the request is loaded asynchronously. */
          disabled={!request}
          /* @end */
          title="Login"
          /* @info Prompt the user to authenticate in a user interaction or web browsers will block it. */
          onPress={promptAsync}
          /* @end */
        />
      )}
    </View>
  );
}
```

Let's concise the example a bit.

**Review by Simek — COMMENTED** [review]
Thanks for the update!  👍 

Left two review comments, at the end decided to comment on whole example at once, since multiple review comments on lines would be more confusing. 

With the suggestions in, the PR should be ready to be shipped. 🙂 

Will also look at the `examples` repo PR next week, thanks for mentioning it!

---

> Few more words on example changes: In general, we should keep those in-doc examples as concise as possible, and avoid noise, so the auth process is the core of it. I saw that you have based this code on the Cognito example, in near future I will compress that one too, and will review other examples on this page.

File: docs/pages/guides/authentication.mdx (hunk 1 above)
Comment (gaokevin1) [reply docs/pages/guides/authentication.mdx:894]: Would it be possible to keep the code exchange part for the token, so customers can copy the whole thing, plug it in, and have it work right away without them having to implement that themselves? The rest is fine to remove.

Comment (Simek) [reply docs/pages/guides/authentication.mdx:894]: The exchange function is still in the code, but it has been moved down to the "success" result code branch, where it's called, there is no need to initialize it earlier.

File: docs/pages/guides/authentication.mdx [hunk 2]
```diff
@@ -770,6 +771,137 @@ function useMounted() {
...
+>
+
+| Website                      | Provider | PKCE      | Auto Discovery |
+| ---------------------------- | -------- | --------- | -------------- |
+| [Get Your Config][c-descope] | OpenID   | Supported | Available      |
+
+[c-descope]: https://docs.descope.com/
+
+- Leverages the Hosted Descope Flow app [Auth-Hosting App](https://github.com/descope/auth-hosting).
+- Requests code after successfully authenticating, followed by exchanging code for the auth tokens (PKCE)
```
Comment (amandeepmittal) [inline docs/pages/guides/authentication.mdx:787]:
```suggestion
- Requests code after successfully authenticating, followed by exchanging code for the auth tokens (PKCE).
```

Missing period.

**Review by Simek — APPROVED** [review]
LGTM, thank you for all of the updates, LGTM! 👌

**Pull request merged**